specs/_features/ePBS/builder.md

```diff
@@ -0,0 +1,229 @@
+        safe_block_hash=safe_block_hash,
+        finalized_block_hash=finalized_block_hash,
+        payload_attributes=payload_attributes,
+    )
+```
+
+2. Set `ExecutionPayload`, where:
+
+```python
+def get_execution_payload(payload_id: Optional[PayloadId], execution_engine: ExecutionEngine) -> ExecutionPayload:
+        return execution_engine.get_payload(payload_id).execution_payload
+```
+
+3. Builder should cache the `ExecutionPayload` and `BlobsBundle` for later use.
+
+4. Convert the `ExecutionPayload` to `ExecutionPayloadHeader` by calling `convert_payload_to_header(payload)`
+
+```python
+def convert_payload_to_header(payload: ExecutionPayload) -> ExecutionPayloadHeader:
+    return ExecutionPayloadHeader(
+        parent_hash=payload.parent_hash,
+        fee_recipient=payload.fee_recipient,
+        state_root=payload.state_root,
+        receipts_root=payload.receipts_root,
+        logs_bloom=payload.logs_bloom,
+        prev_randao=payload.prev_randao,
+        block_number=payload.block_number,
+        gas_limit=payload.gas_limit,
+        gas_used=payload.gas_used,
+        timestamp=payload.timestamp,
+        extra_data=payload.extra_data,
+        base_fee_per_gas=payload.base_fee_per_gas,
+        block_hash=payload.block_hash,
+        transactions_root=hash_tree_root(payload.transactions),
+        withdrawals_root=hash_tree_root(payload.withdrawals),
+        blob_gas_used=payload.blob_gas_used,
+        excess_blob_gas=payload.excess_blob_gas,
+        inclusion_list_summary_root=hash_tree_root(payload.inclusion_list_summary),
+        inclusion_list_exclutions_root=hash_tree_root(payload.inclusion_list_exclutions),
+    )
+```
+5. Build `ExecutionPayloadHeaderEnvelope`
+- Set `header` to the `ExecutionPayloadHeader` from step 3
+- Set `builder_index` to the index of the builder in the beacon state
+- Set `value` to the bid value (ie. how much Gwei builder is willing to pay the proposer if the block is included)
+
+6. Sign the `ExecutionPayloadHeaderEnvelope` with the builder's private key
+
+```python
+def get_execution_payload_header_envelope(state: BeaconState, header: ExecutionPayloadHeaderEnvelope, privkey: int) -> BLSSignature:
+    domain = get_domain(state, DOMAIN_BEACON_BUILDER, compute_epoch_at_slot(state.slot))
+    signing_root = compute_signing_root(header, domain)
+    return bls.Sign(privkey, signing_root)
+```
+
+7. Build `SignedExecutionPayloadHeaderEnvelope`
+- Set `header` to the `ExecutionPayloadHeaderEnvelope` from step 4
+- Set `signature` to the `BLSSignature` from step 5
+
+#### Broadcast execution payload header envelope
+
+Finally, the validator broadcasts `signed_execution_payload_header_envelope` to the global `execution_payload_header_envelope` pubsub topic.
+
+### Signed execution payload envelope construction 
+
+#### Determine if it is safe to reveal
+
+The builder will call `get_head_block`. If the head block contains the payload header, then it is safe to reveal.
```

Comment:
Not sure what the right wording is here. In principle the builder should reveal if he has seen *any* beacon block which is valid and included his payload. Head or not. The problem is that if the call to `get_head_block` returns another block, it will not have his header. If later the attestations switch back to the intended head, the builder will lose his bid.